src/snowflake/connector/network_async.py

```diff
@@ -0,0 +1,637 @@
+) -> None:
+    Error.errorhandler_wrapper(
+        connection,
+        None,
+        DatabaseError,
+        {
+            "msg": f"Failed to get authentication by OKTA: {response.status}: {response.reason}",
+            "errno": ER_FAILED_TO_CONNECT_TO_DB,
+            "sqlstate": SQLSTATE_CONNECTION_REJECTED,
+        },
+    )
+
+
+def raise_failed_request_error_async(
+    connection: SnowflakeConnection | None,
+    url: str,
+    method: str,
+    response: aiohttp.ClientResponse,
+) -> None:
+    # YICHUAN: NO TELEMETRY FOR NOW
+
+    # TelemetryService.get_instance().log_http_request_error(
+    #     f"HttpError{response.status}",
+    #     url,
+    #     method,
+    #     SQLSTATE_CONNECTION_WAS_NOT_ESTABLISHED,
+    #     ER_FAILED_TO_REQUEST,
+    #     response=response,
+    # )
+    Error.errorhandler_wrapper(
+        connection,
+        None,
+        InterfaceError,
+        {
+            "msg": f"{response.status} {response.reason}: {method} {url}",
+            "errno": ER_FAILED_TO_REQUEST,
+            "sqlstate": SQLSTATE_CONNECTION_WAS_NOT_ESTABLISHED,
+        },
+    )
+
+
+# Yichuan: CURRENTLY UNUSED, see SnowflakeRestfulAsync._request_exec_async
+class SnowflakeAuthAsync(aiohttp.BasicAuth):
+    def __new__(cls, token: str) -> aiohttp.BasicAuth:
+        # for class compatibility with BasicAuth, these parameters are never used
+        return super().__new__(cls, "", "", "")
+
+    def __init__(self, token: str) -> None:
+        self.token = token
+
+    def encode(self) -> str:
+        """aiohttp will use this to set the Authorization header"""
+        return HEADER_SNOWFLAKE_TOKEN.format(token=self.token)
+
+
+class SessionPoolAsync(SessionPool):
+    async def close_async(self) -> None:
+        """Closes all active and idle sessions in this session pool."""
+        if self._active_sessions:
+            logger.debug(f"Closing {len(self._active_sessions)} active sessions")
+        for s in itertools.chain(self._active_sessions, self._idle_sessions):
+            try:
+                await s.close()
+            except Exception as e:
+                logger.info(f"Session cleanup failed: {e}")
+        self._active_sessions.clear()
+        self._idle_sessions.clear()
+
+
+class SnowflakeRestfulAsync(SnowflakeRestful):
+    def __init__(
+        self,
+        host: str = "127.0.0.1",
+        port: int = 8080,
+        protocol: str = "http",
+        inject_client_pause: int = 0,
+        connection: SnowflakeConnection | None = None,
+    ) -> None:
+        # YICHUAN: We need to do this instead of using asyncio.run() because of session pooling; if we simply use
+        # asyncio.run(), the loops held by aiohttp.ClientSession will be closed, so the next time the cached session
+        # is used it will raise "RuntimeError: Event loop is closed"
+        self._loop_runner = EventLoopThreadRunner()
+        self._loop_runner.start()
```

Comment:
for referencing: this is for sync api execution, in the future we can move this part to upper stream (e.g. connection.py)